proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,637 @@
+
+Simple rebate schemes will verify merely signers, e.g an oracle preventing 3rd
+parties from write-locking their price feed. Dexes will need more complex rebate
+schemes based on instruction sysvar introspection.
+
+Syscall definition of the rebate mechanism is:
+
+```
+fn rebate_program_rebatable_account_write_fee(
+  address: Pubkey, 
+  microlamports_per_requested_cu: u64
+);
+```
+
+Rebate takes account on which rebate is issued and the amount of microlamports
+per CU that needs to be rebated as input. It returns void but the transaction
+will fail for invalid inputs. In case of multiple rebates on the same account
+only the highest amount of rebate will be taken into account. The rebated amount
+is always the minimum of the largest call to
+`rebate_program_rebatable_account_write_fee()` and the PRAW fees on the
+account. If program rebates `U64::MAX` it means all the PRAW fees on the account
+are rebated. The rebate amount cannot be negative. It will return an error
+`InvalidAccountOwner` if the program calling the syscall is not the owner of the
+account.
+
+### Changes in JSON RPC
+
+Currently JSON RPC service is used to fetch account data from RPC endpoint by
+the clients. We extend the http method `getAccountInfo` to also return the PRAW
+fee in microlamports per CU requested for the account. The http method
+`getFeeForMessage` should be extended to decode the instruction data for the
+program `TransactionHeaderProgram` into `TransactionHeaderParameter`s and then
+calculate the total fees correctly including the PRAW fees. For purpose of
+security we assume that program never rebates the PRAW fees on the accounts.
+Wallets should be extended to use these two methods to display user the correct
+information.
+
+### Calculation of PRAW Fees and rebates.
+
+User creates a transaction by setting maximum PRAW fees in terms of
+microlamports per requested CUs (`m`) and requested CUs (`C`). Solana runtime
+will multiply `m` and `C` to calculate total maximum PRAW fees (`M`) payer is
+willing to pay, then check if user has enough balance to pay all the fees
+including PRAW fees then it will load the accounts required by the transaction.
+
+Then we iterate on accounts to calculate total PRAW fees required (`T`).
+Let (`Pi`) be PRAW fee set for ith account.
+Let number of accounts loaded be (`L`)
+
+```
+T = sum(Pi * C) for i in 0 to L.
+```
+
+If M>=T we continue to execute else the transaction fails with an error
+`ProgramRebatableAccountWriteFeesNotPaid`.
+
+Let Overpaid amount (`O = M - T`)
+
+After execution, we iterate on the rebates issued to calculate total rebated
+amount (`R`), For an account there could be multiple rebates during a
+transaction. Consider that there are `N` rebates on ith account. Each rebate
+will be represented as `Rin`.
+
+```
+R = Sum( min(Pi, max(Rin where n=0 to N)) * C )
+```
+
+We then transfer `(Pi - Ri) * C` into each writable account and `O + R` back to
+the payer.
+
+### Consumption of CUs
```

Comment:
these costs should be measured like the rest of the native ops, not extrapolated

Reply:
Ok noted, will update these costs accordingly once the changes have been done.